# Loader R&D

Objective:
Compare multiple PDF loaders for insurance documents.

Document Used:
SBI_Health_Insurance_Retail.pdf

Metrics:
- Extraction Quality
- Formatting Preservation
- Table Extraction
- Metadata
- Speed

In [8]:
# Uncomment if required

! pip install pymupdf
! pip install pdfplumber
! pip install unstructured

Using cached pymupdf-1.27.2.3-cp310-abi3-win_amd64.whl (19.2 MB)



[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


  Using cached pdfplumber-0.11.9-py3-none-any.whl.metadata (43 kB)
   ---------------------------------------- 0.0/6.6 MB ? eta -:--:--
   - -------------------------------------- 0.3/6.6 MB ? eta -:--:--
   ---- ----------------------------------- 0.8/6.6 MB 2.2 MB/s eta 0:00:03
   ------ --------------------------------- 1.0/6.6 MB 1.9 MB/s eta 0:00:03
   --------- ------------------------------ 1.6/6.6 MB 1.9 MB/s eta 0:00:03
   ------------ --------------------------- 2.1/6.6 MB 2.0 MB/s eta 0:00:03
   -------------- ------------------------- 2.4/6.6 MB 2.0 MB/s eta 0:00:03
   ------------------- -------------------- 3.1/6.6 MB 2.1 MB/s eta 0:00:02
   -------------------- ------------------- 3.4/6.6 MB 2.1 MB/s eta 0:00:02
   ---------------------- ----------------- 3.7/6.6 MB 2.1 MB/s eta 0:00:02
   ------------------------- -------------- 4.2/6.6 MB 2.1 MB/s eta 0:00:02
   ---------------------------- ----------- 4.7/6.6 MB 2.1 MB/s eta 0:00:01
   ------------------------------- 


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
import sys
print(sys.executable)

c:\Users\Vikram\insurance_rag\venv\Scripts\python.exe


In [2]:
! pip install pandas
import time
import pandas as pd

from langchain_community.document_loaders import (
    PyPDFLoader,
    PyMuPDFLoader,
    UnstructuredPDFLoader
)


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


C:\Users\Vikram\AppData\Local\Temp\ipykernel_29984\286518285.py:5: DeprecationWarning: `langchain-community` is being sunset and is no longer actively maintained. See https://github.com/langchain-ai/langchain-community/issues/674 for details and migration guidance toward standalone integration packages.
  from langchain_community.document_loaders import (
c:\Users\Vikram\insurance_rag\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
import sys
print(sys.executable)

c:\Users\Vikram\insurance_rag\venv\Scripts\python.exe


In [4]:
import time

PDF_PATH = "../data/general/health/SBI_Health_Insurance_Retail.pdf"

start = time.time()

loader = PyPDFLoader(PDF_PATH)
docs = loader.load()

end = time.time()

print("Loader: PyPDFLoader")
print("Pages:", len(docs))
print("Time:", round(end - start, 2), "seconds")
print("Metadata:", docs[0].metadata)
print("\nSample Text:\n")
print(docs[0].page_content[:1500])

Loader: PyPDFLoader
Pages: 27
Time: 1.39 seconds
Metadata: {'producer': 'Nitro PDF PrimoPDF', 'creator': 'PrimoPDF http://www.primopdf.com', 'creationdate': '2020-07-16T18:50:21-05:30', 'moddate': '2020-07-16T18:50:21-05:30', 'title': 'Microsoft Word - Health Insurance Policy -Retail', 'author': 'user', 'source': '../data/general/health/SBI_Health_Insurance_Retail.pdf', 'total_pages': 27, 'page': 0, 'page_label': '1'}

Sample Text:

SBI General Insurance Company SBI General Insurance Company SBI General Insurance Company SBI General Insurance Company Limited Limited Limited Limited  
                                                       
 
 
SBI General Insurance Company Limited  SBI General Insurance Company Limited  SBI General Insurance Company Limited  SBI General Insurance Company Limited      
Corporate & Registered Office:   'Natraj', 301, Junction of  Western Express Highway & Andheri - Kurla Road, Andheri (E), Mumbai - 400 069  I 
CIN: U66000MH2009PLC190546  I    Tel.: +91 22

In [5]:
import os

PDF_PATH = "../data/general/health/SBI_Health_Insurance_Retail.pdf"

print(os.path.exists(PDF_PATH))
print(os.path.abspath(PDF_PATH))

True
c:\Users\Vikram\insurance_rag\data\general\health\SBI_Health_Insurance_Retail.pdf


In [6]:
print(repr(docs[0].page_content[:2000]))

"SBI General Insurance Company SBI General Insurance Company SBI General Insurance Company SBI General Insurance Company Limited Limited Limited Limited  \n                                                       \n \n \nSBI General Insurance Company Limited  SBI General Insurance Company Limited  SBI General Insurance Company Limited  SBI General Insurance Company Limited      \nCorporate & Registered Office:   'Natraj', 301, Junction of  Western Express Highway & Andheri - Kurla Road, Andheri (E), Mumbai - 400 069  I \nCIN: U66000MH2009PLC190546  I    Tel.: +91 22 42412000 I   www.sbigeneral.in  I Logo displayed belongs to State Bank of India and is used \nby SBI General Insurance Co. Ltd. under license  I IRDAI Registration Number 144  I  Product Name: Health Insurance Policy Retail. I  \nUIN:  SBIHLIP11002V021011 I IRDAI Reg No.144 \n \nHEALTH INSURANCE POLICY –RETAIL \n \nThis Policy  is issued to the Insured  based on the Proposal  and declaration together with any statement, repor

## PyPDFLoader Evaluation

### Overview

PyPDFLoader was tested on the SBI Health Insurance Retail policy document containing 27 pages.

### Performance

* Total Pages Extracted: 27
* Processing Time: 1.39 seconds

### Text Extraction Quality

* Text was extracted successfully from all pages.
* Policy clauses, headings and insurance terminology were preserved.
* Content was readable without major extraction failures.

### Metadata Extraction

Available metadata:

* Author
* Title
* Source Path
* Total Pages
* Creation Date
* Page Number

### Advantages

* Easy to use.
* Good text extraction quality.
* Suitable for standard insurance policy documents.
* Integrates directly with LangChain pipelines.

### Limitations

* Slower compared to PyMuPDFLoader.
* Repeated company headers appear in extracted text.
* Formatting noise exists in some sections.
* Table structure is not preserved perfectly.

### Conclusion

PyPDFLoader provides reliable extraction for insurance documents and can be used as a baseline loader. However, performance optimization may be required for large document collections.


In [9]:
! pip install pymupdf
start = time.time()

loader = PyMuPDFLoader(PDF_PATH)
docs_mupdf = loader.load()

end = time.time()

print("Loader: PyMuPDFLoader")
print("Pages:", len(docs_mupdf))
print("Time:", round(end - start, 2), "seconds")
print("Metadata:", docs_mupdf[0].metadata)
print("\nSample Text:\n")
print(repr(docs_mupdf[0].page_content[:2000]))


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Loader: PyMuPDFLoader
Pages: 27
Time: 0.54 seconds
Metadata: {'producer': 'Nitro PDF PrimoPDF', 'creator': 'PrimoPDF http://www.primopdf.com', 'creationdate': '2020-07-16T18:50:21-05:30', 'source': '../data/general/health/SBI_Health_Insurance_Retail.pdf', 'file_path': '../data/general/health/SBI_Health_Insurance_Retail.pdf', 'total_pages': 27, 'format': 'PDF 1.3', 'title': 'Microsoft Word - Health Insurance Policy -Retail', 'author': 'user', 'subject': '', 'keywords': '', 'moddate': '2020-07-16T18:50:21-05:30', 'trapped': '', 'modDate': "D:20200716185021-05'30'", 'creationDate': "D:20200716185021-05'30'", 'page': 0}

Sample Text:

"SBI General Insurance Company \nSBI General Insurance Company \nSBI General Insurance Company \nSBI General Insurance Company Limited\nLimited\nLimited\nLimited\n                 \n \nSBI General Insurance Company Limited  \nSBI General Insurance Company Limited  \nSBI General Insurance Company Limited  \nSBI General Insurance Company Limited   \nCorporate &

## PyMuPDFLoader Evaluation

### Overview

PyMuPDFLoader was tested on the SBI Health Insurance Retail policy document containing 27 pages.

### Performance

* Total Pages Extracted: 27
* Processing Time: 0.54 seconds

### Text Extraction Quality

* Text extraction quality was comparable to PyPDFLoader.
* Policy clauses and legal wording were preserved.
* Content remained readable and structured.

### Metadata Extraction

Available metadata:

* Author
* Title
* Source Path
* PDF Format
* Creation Date
* Modification Date
* Total Pages
* Page Number

### Advantages

* Faster than PyPDFLoader.
* Richer metadata extraction.
* Suitable for large-scale ingestion pipelines.
* Good compatibility with insurance policy documents.

### Limitations

* Repeated headers still appear.
* Some formatting artifacts remain.
* Table structures are not fully preserved.

### Comparison with PyPDFLoader

* Similar text quality.
* Better speed.
* More detailed metadata.
* Better suited for large document repositories.

### Conclusion

PyMuPDFLoader achieved the best balance between extraction quality, metadata availability, and processing speed. Based on the current experiment, it is the preferred loader for insurance policy documents.


In [10]:
! pip install unstructured


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [11]:
from langchain_community.document_loaders import UnstructuredPDFLoader
import time

In [13]:
! pip install pi-heif
PDF_PATH = "../data/general/health/SBI_Health_Insurance_Retail.pdf"

start = time.time()

loader = UnstructuredPDFLoader(PDF_PATH)
docs_unstructured = loader.load()

end = time.time()

print("Loader: UnstructuredPDFLoader")
print("Pages/Elements:", len(docs_unstructured))
print("Time:", round(end - start, 2), "seconds")

print("\nMetadata:\n")
print(docs_unstructured[0].metadata)

print("\nSample Text:\n")
print(repr(docs_unstructured[0].page_content[:2000]))

   ---------------------------------------- 0.0/1.9 MB ? eta -:--:--
   ----- ---------------------------------- 0.3/1.9 MB ? eta -:--:--
   ---------- ----------------------------- 0.5/1.9 MB 2.2 MB/s eta 0:00:01
   ---------------- ----------------------- 0.8/1.9 MB 2.2 MB/s eta 0:00:01
   -------------------------- ------------- 1.3/1.9 MB 1.8 MB/s eta 0:00:01
   -------------------------------- ------- 1.6/1.9 MB 1.7 MB/s eta 0:00:01
   ---------------------------------------- 1.9/1.9 MB 1.7 MB/s eta 0:00:00



[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


ModuleNotFoundError: No module named 'unstructured_inference'

In [14]:
! pip install unstructured-inference

   ---------------------------------------- 0.0/8.2 MB ? eta -:--:--
   - -------------------------------------- 0.3/8.2 MB ? eta -:--:--
   --- ------------------------------------ 0.8/8.2 MB 2.3 MB/s eta 0:00:04
   ------ --------------------------------- 1.3/8.2 MB 2.3 MB/s eta 0:00:03
   -------- ------------------------------- 1.8/8.2 MB 2.3 MB/s eta 0:00:03
   ----------- ---------------------------- 2.4/8.2 MB 2.3 MB/s eta 0:00:03
   ------------ --------------------------- 2.6/8.2 MB 2.4 MB/s eta 0:00:03
   --------------- ------------------------ 3.1/8.2 MB 2.3 MB/s eta 0:00:03
   ----------------- ---------------------- 3.7/8.2 MB 2.3 MB/s eta 0:00:03
   ------------------- -------------------- 3.9/8.2 MB 2.2 MB/s eta 0:00:02
   -------------------- ------------------- 4.2/8.2 MB 2.2 MB/s eta 0:00:02
   ---------------------- ----------------- 4.7/8.2 MB 2.2 MB/s eta 0:00:02
   ------------------------- -------------- 5.2/8.2 MB 2.2 MB/s eta 0:00:02
   -----------------------


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


UnstructuredPDFLoader requires heavier dependency setup compared to PyPDFLoader and PyMuPDFLoader. During testing, missing dependencies like pi_heif and unstructured_inference were required.

In [15]:
! pip install pdfplumber


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [16]:
import pdfplumber
import time

PDF_PATH = "../data/general/health/SBI_Health_Insurance_Retail.pdf"

start = time.time()

plumber_pages = []

with pdfplumber.open(PDF_PATH) as pdf:
    for i, page in enumerate(pdf.pages):
        text = page.extract_text()
        tables = page.extract_tables()
        plumber_pages.append({
            "page": i,
            "text": text,
            "tables": tables
        })

end = time.time()

print("Loader: PDFPlumber")
print("Pages:", len(plumber_pages))
print("Time:", round(end - start, 2), "seconds")
print("Tables on first page:", len(plumber_pages[0]["tables"]))
print("\nSample Text:\n")
print(repr(plumber_pages[0]["text"][:2000] if plumber_pages[0]["text"] else ""))

Loader: PDFPlumber
Pages: 27
Time: 4.01 seconds
Tables on first page: 0

Sample Text:

'SSSSBBBBIIII GGGGeeeennnneeeerrrraaaallll IIIInnnnssssuuuurrrraaaannnncccceeee CCCCoooommmmppppaaaannnnyyyy LLLLiiiimmmmiiiitttteeeedddd\nHEALTH INSURANCE POLICY –RETAIL\nThis Policy is issued to the Insured based on the Proposal and declaration together with any statement, report or\nother document which shall be the basis of this contract and shall be deemed to be incorporated herein, to Insurer\nupon payment of the Premium. This Policy records the agreement between Insurer and Insured and sets out the\nterms of insurance and the obligations of each party.\nThe Policy, the Schedule and any Endorsement shall be read together and any word or expression to which a\nspecific meaning has been attached in any part of this Policy or of Schedule shall bear such meaning whenever\nit may appear.\nSubject to the terms, Conditions, exclusions and definitions contained herein or endorsed or otherwise\nexpresse

## PDFPlumber Evaluation

### Overview
PDFPlumber was tested on the SBI Health Insurance Retail policy document containing 27 pages.

### Performance
- Total Pages Extracted: 27
- Processing Time: 4.01 seconds
- Tables Detected on First Page: 0

### Text Extraction Quality
- Text was extracted, but the first line showed repeated characters.
- Example issue: `SSSSBBBBIIII GGGGeeeennnneeeerrrraaaallll`
- Main policy text was readable after the noisy header.
- Bullet points were extracted, but special bullet symbols appeared as unicode characters.

### Table Extraction
- No tables were detected on the first page.
- For this specific PDF, PDFPlumber did not provide a table extraction advantage.

### Advantages
- Can be useful for PDFs with clear table structures.
- Allows page-level extraction and table extraction separately.
- Useful for deeper table-specific experiments.

### Limitations
- Slowest among tested loaders so far.
- Header text extraction was noisy.
- No table detected on the first page.
- Extra cleaning would be required before using this output in RAG.

### Conclusion
PDFPlumber is not preferred as the default loader for this SBI Health Insurance PDF because it was slower and introduced character duplication noise. It may still be useful as a fallback for PDFs where table extraction is important.

# Loader Comparison

| Feature | PyPDFLoader | PyMuPDFLoader | PDFPlumber |
|----------|----------|----------|----------|
| Pages Extracted | 27 | 27 | 27 |
| Time | 1.39 sec | 0.54 sec | 4.01 sec |
| Text Quality | Good | Good | Moderate |
| Metadata | Good | Excellent | Limited |
| Table Extraction | Limited | Limited | No advantage found |
| Setup Complexity | Easy | Easy | Easy |
| Cleaning Required | Medium | Medium | High |
| RAG Suitability | High | Very High | Medium |

## Conclusion

✅ **PyMuPDFLoader**
- Fastest
- Rich metadata
- Good text extraction

✅ **PyPDFLoader**
- Reliable fallback
- Good extraction quality

⚠️ **PDFPlumber**
- Slowest
- Character duplication issues observed
- Better suited for table-heavy PDFs

### Selected Loader

**Primary:** PyMuPDFLoader  
**Fallback:** PyPDFLoader  
**Special Case:** PDFPlumber (table-heavy documents)